zeus/device/soc/jetson.py

```diff
@@ -0,0 +1,272 @@
+
+import abc
+
+from typing import Set
+from pathlib import Path
+import enum
+from dataclasses import dataclass
+
+import multiprocessing as mp
+import time
+import atexit
+import asyncio
+
+import zeus.device.soc.common as soc_common
+
+
+def check_file(path):
+    """Check if the given path exists and is a file."""
+    path = Path(path)
+    return path.exists() and path.is_file()
+
+
+class PowerMeasurementStrategy(abc.ABC):
+    """Abstract base class for two different power measurement strategies."""
+
+    @abc.abstractmethod
+    def measure_power(self) -> float:
+        """Measure power in mW."""
+        pass
+
+
+class DirectPower(PowerMeasurementStrategy):
+    """Reads power directly from a sysfs path."""
+
+    def __init__(self, power_path: Path):
+        """Initialize DirectPower paths."""
+        self.power_path = power_path
+
+    def measure_power(self) -> float:
+        """Measure power by reading from sysfs paths.
+
+        Units: mW.
+        """
+        power = float(self.power_path.read_text().strip())
+        return power / 1000
+
+
+class VoltageCurrentProduct(PowerMeasurementStrategy):
+    """Computes power as product of voltage and current, read from two sysfs paths."""
+
+    def __init__(self, voltage_path: Path, current_path: Path):
+        """Initialize VoltageCurrentProduct paths."""
+        self.voltage_path = voltage_path
+        self.current_path = current_path
+
+    def measure_power(self) -> float:
+        """Measure power by reading from sysfs paths.
+
+        Units: mW.
+        """
+        voltage = float(self.voltage_path.read_text().strip())
+        current = float(self.current_path.read_text().strip())
+        return (voltage * current) / 1000
+
+
+@dataclass
+class JetsonMeasurement(soc_common.SoCMeasurement):
+    """Represents energy measurements for Jetson Nano subsystems."""
+
+    cpu_energy_mj: float = 0.0
+    gpu_energy_mj: float = 0.0
+    total_energy_mj: float = 0.0
+
+    def __sub__(self, other: JetsonMeasurement) -> JetsonMeasurement:
+        """Return a new JetsonMeasurement with subtracted field values."""
+        return JetsonMeasurement(
+            cpu_energy_mj=self.cpu_energy_mj - other.cpu_energy_mj,
+            gpu_energy_mj=self.gpu_energy_mj - other.gpu_energy_mj,
+            total_energy_mj=self.total_energy_mj - other.total_energy_mj,
+        )
+
+    def zeroAllFields(self) -> None:
+        """Set all internal measurement values to zero."""
+        self.cpu_energy_mj = 0.0
+        self.gpu_energy_mj = 0.0
+        self.total_energy_mj = 0.0
+
+
+class Jetson(soc_common.SoC):
+    """An interface for obtaining the energy metrics of a Jetson Nano processor."""
+
+    def __init__(self) -> None:
+        """Initialize Jetson monitoring object."""
+        super().__init__()
+        self.metric_paths = self._discover_metrics_and_paths()
+        self.power_measurement = {}  # maps rail to PowerMeasurementStrategy object
```

Comment:
Many variables, functions, function arguments, and function return values are missing their type annotations.

Especially for this one, what are the possible values for the keys? cpu, gpu, and total?